Add DynamicFacingInit to change the facing of previews dynamically

teees opened the pull request:
Needed for carryalls (#10775)

teees committed:
Add DynamicFacingInit to change the facing of previews dynamically

FIXUP More render traits using dynamic facing

FIXUP Use functions within SpriteActorPreview to get dynamic offsets for turrets

pchote commented on OpenRA.Game/Map/ActorInitializer.cs:
```diff
@@ -53,7 +54,23 @@ public class FacingInit : IActorInit<int>
+		public override int Value(World world) { return func(); }
+	}
+
+	public class BodyAnimationFrameInit : IActorInit<uint>
```
It looks like all of these `*Init`s except `OwnerInit` can be moved into mod code.  Can you create an `ActorInitializers.cs` file in Mods.Common for them?

teees replied:
Moved all of them to Common, except OwnerInit and LocationInit

pchote commented on OpenRA.Mods.Common/Traits/Render/WithIdleOverlay.cs:
```diff
@@ -47,13 +48,13 @@ public IEnumerable<IActorPreview> RenderPreviewSprites(ActorPreviewInitializer i
-			var orientation = body.QuantizeOrientation(new WRot(WAngle.Zero, WAngle.Zero, WAngle.FromFacing(facing)), facings);
-			var offset = body.LocalToWorld(Offset.Rotate(orientation));
-			yield return new SpriteActorPreview(anim, offset, offset.Y + offset.Z + 1, p, rs.Scale);
+			Func<WRot> orientation = () => body.QuantizeOrientation(new WRot(WAngle.Zero, WAngle.Zero, WAngle.FromFacing(facing())), facings);
```
You can simplify this down to `Func<WRot> orientation = () => body.QuantizeOrientation(WRot.FromFacing(facing()), facings);`

pchote commented on OpenRA.Mods.Common/Traits/Render/WithIdleOverlay.cs:
```diff
@@ -47,13 +48,13 @@ public IEnumerable<IActorPreview> RenderPreviewSprites(ActorPreviewInitializer i
-			yield return new SpriteActorPreview(anim, offset, offset.Y + offset.Z + 1, p, rs.Scale);
+			Func<WRot> orientation = () => body.QuantizeOrientation(new WRot(WAngle.Zero, WAngle.Zero, WAngle.FromFacing(facing())), facings);
+			Func<WVec> offset = () => body.LocalToWorld(Offset.Rotate(orientation()));
+			yield return new SpriteActorPreview(anim, offset, () => offset().Y + offset().Z + 1, p, rs.Scale);
```
Kind of sucks to evaluate `offset` twice.  Could you pull out an `Action<int>` that calculates it once and then takes the `.Y` and `.Z` from the result?

pchote commented on OpenRA.Mods.Common/Traits/Render/WithInfantryBody.cs:
```diff
@@ -31,14 +31,10 @@ public class WithInfantryBodyInfo : UpgradableTraitInfo, IRenderActorPreviewSpri
-				facing = init.Contains<FacingInit>() ? init.Get<FacingInit, int>() : ifacing.GetInitialFacing();
-
-			var anim = new Animation(init.World, image, () => facing);
+			var anim = new Animation(init.World, image, () => ifacing != null ? init.Contains<FacingInit>() ? init.Get<FacingInit, int>() : ifacing.GetInitialFacing() : 0);
```
The `init` and `GetInitialFacing` aren't going to change, so moving them here hurts perf without giving us anything.  Can we keep this as it was?

pchote replied:
Actually this comments apply to most of the uses of `init.Contains<FacingInit>()`.  Hiding the dynamic-ness of the facing inside `DynamicFacingInit.Value` makes it easy to mess up the use of facings.  I think it would be much better to pull out a helper method that takes an init dict, a fallback facing (from InitialFacing or elsewhere), and returns a `Func<int>`.  Internally it would explicitly check for `DynamicFacingInit` vs `FacingInit`.

pchote commented on OpenRA.Mods.Common/Traits/Render/WithParachute.cs:
```diff
@@ -73,9 +74,9 @@ public IEnumerable<IActorPreview> RenderPreviewSprites(ActorPreviewInitializer i
 
 			var body = init.Actor.TraitInfo<BodyOrientationInfo>();
 			var facing = init.Contains<FacingInit>() ? init.Get<FacingInit, int>() : 0;
```
This is going to miss the dynamic facing changes.

pchote commented on OpenRA.Mods.Common/Traits/Render/WithSpriteBarrel.cs:
```diff
@@ -46,10 +47,10 @@ public IEnumerable<IActorPreview> RenderPreviewSprites(ActorPreviewInitializer i
 
-			var turretOrientation = body.QuantizeOrientation(new WRot(WAngle.Zero, WAngle.Zero, WAngle.FromFacing(t.InitialFacing)), facings);
-			var turretOffset = body.LocalToWorld(t.Offset.Rotate(turretOrientation));
+			Func<WRot> turretOrientation = () => body.QuantizeOrientation(new WRot(WAngle.Zero, WAngle.Zero, WAngle.FromFacing(t.InitialFacing)), facings);
```
Can use `WRot.FromFacing` here too.

pchote replied:
The turret facing doesn't change, so this can stay as an `int` that you wrap in a `Func` in the ctor below.

pchote commented on OpenRA.Mods.Common/Traits/Render/WithSpriteRotorOverlay.cs:
```diff
@@ -33,13 +34,13 @@ public class WithSpriteRotorOverlayInfo : ITraitInfo, IRenderActorPreviewSprites
-			var orientation = body.QuantizeOrientation(new WRot(WAngle.Zero, WAngle.Zero, WAngle.FromFacing(facing)), facings);
-			var offset = body.LocalToWorld(Offset.Rotate(orientation));
-			yield return new SpriteActorPreview(anim, offset, offset.Y + offset.Z + 1, p, rs.Scale);
+			Func<WRot> orientation = () => body.QuantizeOrientation(new WRot(WAngle.Zero, WAngle.Zero, WAngle.FromFacing(facing())), facings);
```
`WRot.FromFacing` (note that this helper was added after all this code was originally written, and I apparently forgot to backport it).

pchote commented on OpenRA.Mods.Common/Traits/Render/WithSpriteRotorOverlay.cs:
```diff
@@ -33,13 +34,13 @@ public class WithSpriteRotorOverlayInfo : ITraitInfo, IRenderActorPreviewSprites
-			yield return new SpriteActorPreview(anim, offset, offset.Y + offset.Z + 1, p, rs.Scale);
+			Func<WRot> orientation = () => body.QuantizeOrientation(new WRot(WAngle.Zero, WAngle.Zero, WAngle.FromFacing(facing())), facings);
+			Func<WVec> offset = () => body.LocalToWorld(Offset.Rotate(orientation()));
+			yield return new SpriteActorPreview(anim, offset, () => offset().Y + offset().Z + 1, p, rs.Scale);
```
We want to avoid evaluating `offset()` twice here, too.

pchote commented on OpenRA.Mods.Common/Traits/Render/WithSpriteTurret.cs:
```diff
@@ -45,15 +46,15 @@ public IEnumerable<IActorPreview> RenderPreviewSprites(ActorPreviewInitializer i
-			yield return new SpriteActorPreview(anim, offset, -(offset.Y + offset.Z) + 1, p, rs.Scale);
+			Func<WRot> orientation = () => body.QuantizeOrientation(new WRot(WAngle.Zero, WAngle.Zero, WAngle.FromFacing(bodyFacing())), facings);
+			Func<WVec> offset = () => body.LocalToWorld(t.Offset.Rotate(orientation()));
+			yield return new SpriteActorPreview(anim, offset, () => -(offset().Y + offset().Z) + 1, p, rs.Scale);
```
Here too.

pchote commented on OpenRA.Mods.Common/Traits/Render/WithVoxelTurret.cs:
```diff
@@ -37,12 +39,12 @@ public IEnumerable<VoxelAnimation> RenderPreviewVoxels(ActorPreviewInitializer i
-			var turretBodyOrientation = new WRot(WAngle.Zero, WAngle.Zero, WAngle.FromFacing(turretFacing) - orientation.Yaw);
-			var turretOrientation = new[] { turretBodyOrientation, body.QuantizeOrientation(orientation, facings) };
-			yield return new VoxelAnimation(voxel, () => turretOffset, () => turretOrientation, () => false, () => 0);
+			Func<WRot> turretBodyOrientation = () => new WRot(WAngle.Zero, WAngle.Zero, WAngle.FromFacing(turretFacing) - orientation().Yaw);
```
Can use `WRot.FromYaw` here (`FromFacing` won't work).

pchote commented on OpenRA.Mods.Common/TraitsInterfaces.cs:
```diff
@@ -124,4 +124,9 @@ interface IWallConnector
+
+	public interface IBodyAnimation
+	{
+		uint GetBodyAnimationFrame();
```
Might be cleaner as a property instead?

Phrohdoh replied:
Shouldn't this take a `self` param and as such remain a method for consistency?

pchote commented on OpenRA.Mods.Common/OpenRA.Mods.Common.csproj:
```diff
@@ -753,4 +754,4 @@ cd "$(SolutionDir)"</PostBuildEvent>
   </Target>
   -->
-</Project>
+</Project>
```
Please keep the EOF newline here (just drop this line change).

pchote commented on OpenRA.Mods.Common/Traits/Render/RenderVoxels.cs:
```diff
@@ -59,8 +60,18 @@ public virtual IEnumerable<IActorPreview> RenderPreview(ActorPreviewInitializer
-			var facing = ifacing != null ? init.Contains<FacingInit>() ? init.Get<FacingInit, int>() : ifacing.GetInitialFacing() : 0;
-			var orientation = WRot.FromFacing(facing);
+			Func<WRot> orientation = () => WRot.Zero;
+			if (ifacing != null)
```
This block is duplicated across a few places.  Can you pull it out into a helper method?

pchote commented on OpenRA.Mods.Common/Traits/Render/WithFacingSpriteBody.cs:
```diff
@@ -23,12 +24,22 @@ public class WithFacingSpriteBodyInfo : WithSpriteBodyInfo, Requires<BodyOrienta
 			var ifacing = init.Actor.TraitInfoOrDefault<IFacingInfo>();
-			var facing = ifacing != null ? init.Contains<FacingInit>() ? init.Get<FacingInit, int>() : ifacing.GetInitialFacing() : 0;
+			Func<int> facing = () => 0;
+			if (ifacing != null)
```
Likewise for this (for facing instead of WRot).

pchote commented on OpenRA.Mods.Common/Traits/Render/WithIdleOverlay.cs:
```diff
@@ -47,13 +48,19 @@ public IEnumerable<IActorPreview> RenderPreviewSprites(ActorPreviewInitializer i
 			var body = init.Actor.TraitInfo<BodyOrientationInfo>();
-			var facing = init.Contains<FacingInit>() ? init.Get<FacingInit, int>() : 0;
-			var anim = new Animation(init.World, image, () => facing);
+			Func<int> facing = init.Contains<FacingInit>() ? (Func<int>)(() => init.Get<FacingInit, int>()) : () => 0;
```
Do you really need the inner `(Func<int>)` cast here?

teees replied:
I could put it in yet another helper function - it's a bit tricky to assign a lambda function using the ternary operator. (see http://stackoverflow.com/questions/11309557/assign-a-lambda-expression-using-the-conditional-ternary-operator)

pchote commented on OpenRA.Mods.Common/Traits/Render/WithVoxelBody.cs:
```diff
@@ -25,13 +25,13 @@ public class WithVoxelBodyInfo : UpgradableTraitInfo, IRenderActorPreviewVoxelsI
-			var bodyOrientation = new[] { body.QuantizeOrientation(orientation, facings) };
 			yield return new VoxelAnimation(voxel, () => WVec.Zero,
-				() => bodyOrientation,
+				() => new[] { body.QuantizeOrientation(orientation(), facings) },
```
@RoosterDragon would we win anything noticable by reusing the array (updating its value) instead of creating new objects in cases like this?

RoosterDragon replied:
I wouldn't worry about it much unless it showed up in profiling as expensive.

teees commented:
Rebased

pchote commented on OpenRA.Mods.Common/ActorInitializer.cs:
```diff
@@ -0,0 +1,60 @@
+		public virtual int Value(World world) { return value; }
+	}
+
+	public class DynamicFacingInit : FacingInit
```
I mentioned this earlier, but it may have been lost in the scrum of other comments: can we please make this not subclass `FacingInit`?  My suggestions for the `GetFacing` and `GetOrientation` methods were specifically to enable this separation :smile:

The issue is that ActorInits are supposed to be immutable, and while this technically doesn't change that it means that in practice you can query something that you *think* is an immutable `FacingInit`, but is actually an immutable `DynamicFacingInit` that gives you different facing values each time you query it – violating expectations for a regular `FacingInit`.

teees commented:
@pchote I separated those two cases as far as possible, there is some code duplication, but it should probably save some resources. Not sure if I could get those into some helper function.

Also rebased, and rebased #10775 on top of it

teees commented:
Read your comment again and made the dynamicfacing actor initializer return a `Func<int>` as requested

teees commented:
rebased

pchote commented:
When testing this with #10775 actor turrets don't track the dynamic facing.  This is most visible with the Titan.

pchote commented:
Code looks good, and didn't spot any other issues aside from that.

RoosterDragon commented on OpenRA.Mods.Common/Traits/Render/WithIdleOverlay.cs:
```diff
@@ -47,13 +48,34 @@ public IEnumerable<IActorPreview> RenderPreviewSprites(ActorPreviewInitializer i
+				var offset = body.LocalToWorld(Offset.Rotate(orientation));
+				var zOffset = offset.Y + offset.Z + 1;
+				yield return new SpriteActorPreview(anim, () => offset, () => zOffset, p, rs.Scale);
+			}
```
Is there a way to extract this code into a common method somewhere? It's far from trivial and used in several places.

pchote commented:
@teees: would you mind if someone else took this over to do the final fixes so we can get it merged?

teees commented:
@pchote of course no, feel free to take over anything needed :)

pchote commented:
I'll take this over and file a new PR when ready.

pchote commented:
While looking into the turret facing issue I found a problem that is going to keep on causing trouble if we don't fix it now.  I'll fix that and include my old rework of cargo rendering (which is finally feasible with this PR) as a testcase.  This will take a bit longer than I first expected.
